cloud-controller-manager/do/firewall_controller_test.go

```diff
@@ -0,0 +1,456 @@
+	inboundRule := godo.InboundRule{
+		Protocol:  "tcp",
+		PortRange: "31220",
+	}
+	nodePortService := &v1.Service{
+		Spec: v1.ServiceSpec{
+			Type: v1.ServiceTypeNodePort,
+			Ports: []v1.ServicePort{
+				{
+					Protocol: "tcp",
+					Port:     31220,
+					NodePort: 31220,
+				},
+			},
+		},
+	}
+	fakeServiceList := []*v1.Service{nodePortService}
+
+	testcases := []struct {
+		name        string
+		firewall    *godo.Firewall
+		fwCache     firewallCache
+		serviceList []*v1.Service
+	}{
+		{
+			name:        "successfully updates port range",
+			firewall:    nil,
+			fwCache:     newFakeFirewallCache(inboundRule),
+			serviceList: fakeServiceList,
+		},
+		{
+			name:        "return node port inbound rules when we find a service of type nodeport",
+			firewall:    newFakeFirewall(fakeWorkerFWName, inboundRule),
+			fwCache:     newFakeFirewallCache(inboundRule),
+			serviceList: fakeServiceList,
+		},
+	}
+
+	for _, test := range testcases {
+		t.Run(test.name, func(t *testing.T) {
+			// setup
+			gclient := newFakeGodoClient(
+				&fakeFirewallService{},
+			)
+			inboundRules := []godo.InboundRule{inboundRule}
+			fwManagerOp = newFakeFirewallManagerOp(gclient, test.fwCache)
+			fc := NewFirewallController(ctx, kclient, gclient, inf.Core().V1().Services(), fwManagerOp, []string{})
+
+			rules := fc.createInboundRules(test.serviceList)
+			if !cmp.Equal(rules[0].PortRange, inboundRules[0].PortRange) {
+				t.Errorf("got %q, want %q", rules[0].PortRange, inboundRules[0].PortRange)
+			}
+			if !cmp.Equal(rules[0].Protocol, inboundRules[0].Protocol) {
+				t.Errorf("got %q, want %q", rules[0].Protocol, inboundRules[0].Protocol)
+			}
+		})
+	}
+}
+
+func TestFirewallController_NoDataRace(t *testing.T) {
+	// setup
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	defer cancel()
+	var wg sync.WaitGroup
+
+	gclient := newFakeGodoClient(
+		&fakeFirewallService{
+			listFunc: func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error) {
+				return nil, newFakeOKResponse(), nil
+			},
+			updateFunc: func(context.Context, string, *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error) {
+				return newFakeFirewall(fakeWorkerFWName, fakeInboundRule), newFakeOKResponse(), nil
+			},
+			createFunc: func(context.Context, *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error) {
+				return nil, newFakeOKResponse(), nil
+			},
+			getFunc: func(context.Context, string) (*godo.Firewall, *godo.Response, error) {
+				return newFakeFirewall(fakeWorkerFWName, fakeInboundRule), newFakeOKResponse(), nil
+			},
+		},
+	)
+	fwManagerOp := newFakeFirewallManagerOp(gclient, newFakeFirewallCache(fakeInboundRule))
+	fc := NewFirewallController(ctx, kclient, gclient, inf.Core().V1().Services(), fwManagerOp, []string{})
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for ctx.Err() == nil { // context has not been terminated
+			fc.ensureReconciledFirewall(ctx)
```

Comment:
Just realized we should probably handle the error returned from `ensureReconciledFirewall`, only ignoring it if it's about the canceled context:

```go
if err := fc.ensureReconciledFirewall(ctx); err != context.Canceled {
  t.Errorf("ensureReconciledFirewall failed: %s", err)
}
```